nomt/src/beatree/ops/update/branch_updater.rs

```diff
@@ -498,4 +502,378 @@ mod tests {
+            }),
+            None,
+        );
+        let mut branch_changes = BranchChanges::default();
+
+        updater.ingest(key(251), Some(9999.into()));
+        let DigestResult::Finished = updater.digest(&mut branch_changes) else {
+            panic!()
+        };
+
+        let new_branch_entry = branch_changes.get(key(0)).unwrap();
+
+        let new_branch = new_branch_entry.inserted.as_ref().unwrap();
+        assert_eq!(new_branch.n(), 501);
+        assert_eq!(new_branch.node_pointer(0), 0);
+        assert_eq!(new_branch.node_pointer(500), 499);
+        assert_eq!(new_branch.node_pointer(126), 9999);
+    }
+
+    #[test]
+    fn insert_overflowing() {
+        let key = |i| prefixed_key(0x11, 5, i);
+        let branch = make_branch_with_body_size_target(key, |size| size <= BRANCH_NODE_BODY_SIZE);
+        let n = branch.n() as usize;
+
+        let mut updater = BranchUpdater::new(
+            PAGE_POOL.clone(),
+            Some(BaseBranch {
+                node: branch,
+                iter_pos: 0,
+            }),
+            None,
+        );
+        let mut branch_changes = BranchChanges::default();
+
+        updater.ingest(key(n), Some(PageNumber(n as u32)));
+        let DigestResult::Finished = updater.digest(&mut branch_changes) else {
+            panic!()
+        };
+
+        let new_branch_entry_1 = branch_changes.get(key(0)).unwrap();
+        let new_branch_1 = new_branch_entry_1.inserted.as_ref().unwrap();
+
+        let new_branch_entry_2 = branch_changes.get(key(new_branch_1.n() as usize)).unwrap();
+        let new_branch_2 = new_branch_entry_2.inserted.as_ref().unwrap();
+
+        assert_eq!(new_branch_1.node_pointer(0), 0);
+        assert_eq!(
+            new_branch_2.node_pointer((new_branch_2.n() - 1) as usize),
+            n as u32
+        );
+    }
+
+    #[test]
+    fn delete() {
+        let key = |i| prefixed_key(0x11, 5, i);
+        let branch = make_branch((0..500).map(|i| (key(i), i)).collect());
+
+        let mut updater = BranchUpdater::new(
+            PAGE_POOL.clone(),
+            Some(BaseBranch {
+                node: branch,
+                iter_pos: 0,
+            }),
+            None,
+        );
+        let mut branch_changes = BranchChanges::default();
+
+        updater.ingest(key(250), None);
+        let DigestResult::Finished = updater.digest(&mut branch_changes) else {
+            panic!()
+        };
+
+        let new_branch_entry = branch_changes.get(key(0)).unwrap();
+
+        let new_branch = new_branch_entry.inserted.as_ref().unwrap();
+        assert_eq!(new_branch.n(), 499);
+        assert_eq!(new_branch.node_pointer(0), 0);
+        assert_eq!(new_branch.node_pointer(498), 499);
+    }
+
+    #[test]
+    fn delete_underflow_and_merge() {
+        let key = |i| prefixed_key(0xff, 5, i);
+        let key2 = |i| prefixed_key(0xff, 6, i);
+
+        let mut rightsized = false;
+        let branch = make_branch_with_body_size_target(key, |size| {
```

Comment:
I see that `rightsized` is being used because `make_branch_with_body_size_target` uses `gauge.body_size_after` to stop the loop. Therefore, you need to perform one more iteration to reach the desired node state for testing. A simpler approach might be to use `gauge.body_size`, move at the end of the loop the condition to stop it and then use `let branch = make_branch_with_body_size_target(key, |size| size < BRANCH_MERGE_THRESHOLD);` to achieve the same branch size target

However, the proposed solution using 'body_size_after' may have been intentional to test it implicitly

Reply:
> you need to perform one more iteration to reach the desired node state for testing

The code still looks correct to me as-is. The predicate returns true with the first item that surpasses the merge threshold and false for every item afterwards.

Reply:
Oh, yes, I never intended to say your code was wrong. I just wanted to provide a possibly simpler approach